groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3548 @@
+};
+
+template <class TYPE, class... OTHER>
+struct Variant_UsesBslmaAllocatorAny
+: bsl::integral_constant<
+      bool,
+      BloombergLP::bslma::UsesBslmaAllocator<TYPE>::value ||
+          BloombergLP::bslstl::Variant_UsesBslmaAllocatorAny<
+              OTHER...>::value> {
+};
+
+// This metafunction is derived from 'bsl::true_type' if all template arguments
+// are bitwise moveable, and from 'bsl::false_type' otherwise.  This
+// metafunction is component-private.
+template <class TYPE = VARIANT_NOT_A_TYPE, class... OTHER>
+struct Variant_IsBitwiseMoveableAll;
+
+template <>
+struct Variant_IsBitwiseMoveableAll<VARIANT_NOT_A_TYPE> : bsl::false_type {
+};
+
+template <class TYPE, class... OTHER>
+struct Variant_IsBitwiseMoveableAll
+: bsl::integral_constant<bool,
+                         BloombergLP::bslmf::IsBitwiseMoveable<TYPE>::value &&
+                             Variant_IsBitwiseMoveableAll<OTHER...>::value> {
+};
+#endif
+
+// This metafunction is derived from 'bsl::true_type' if 'T' is not a tag type.
+// This metafunction requires any cv and ref qualifications to be removed from
+// the queried type.
+template <typename TYPE>
+struct Variant_IsTag : bsl::false_type {
+};
+
+template <>
+struct Variant_IsTag<bsl::allocator_arg_t> : bsl::true_type {
+};
+
+template <typename TYPE>
+struct Variant_IsTag<bsl::in_place_type_t<TYPE> > : bsl::true_type {
+};
+
+template <size_t INDEX>
+struct Variant_IsTag<bsl::in_place_index_t<INDEX> > : bsl::true_type {
+};
+
+#define BSLSTL_VARIANT_DEFINE_IF_CONSTRUCTS_FROM(VARIANT, TYPE)               \
+    typename bsl::enable_if<                                                  \
+        BloombergLP::bslstl::Variant_ConstructsFromType<VARIANT,              \
+                                                        TYPE>::value,         \
+        BloombergLP::bslstl::Variant_NoSuchType>::type
+
+#define BSLSTL_VARIANT_DECLARE_IF_CONSTRUCTS_FROM(VARIANT, TYPE)              \
+    BSLSTL_VARIANT_DEFINE_IF_CONSTRUCTS_FROM(                                 \
+                        VARIANT,                                              \
+                        TYPE) = BloombergLP::bslstl::Variant_NoSuchType(0)
+
+#define BSLSTL_VARIANT_INDEX_OF(TYPE, VARIANT)                                \
+    BloombergLP::bslstl::Variant_TypeToIndex<TYPE, VARIANT>::value
+
+#define BSLSTL_VARIANT_CONVERT_INDEX_OF(TYPE, VARIANT)                        \
+    BloombergLP::bslstl::Variant_ConvertIndex<TYPE, VARIANT>::value
+
+#ifdef BSL_VARIANT_FULL_IMPLEMENTATION
+#define BSLSTL_VARIANT_VISITID(RET, VARIANT, VISITOR, VAROBJ)                 \
+    BloombergLP::bslstl::VisitImplId<RET, bsl::variant_size<VARIANT>::value>( \
+        VISITOR, VAROBJ);
+#else
+#define BSLSTL_VARIANT_VISITID(RET, VARIANT, VISITOR, VAROBJ)                 \
+    BloombergLP::bslstl::                                                     \
+                        VisitImplId<RET>(VISITOR, VAROBJ);
```

Comment:
So my opinion&mdash;and feel free to take it with a grain of salt&mdash;is that if you pass the variant size explicitly when instantiating `VisitImplId`, and especially when you only do it in the full implementation, you give the impression that that template parameter is actually needed, possibly because there might be cases elsewhere in which you might instantiate `VisitImplId` with a size parameter that's different from the actual number of alternatives in the variant type. So I do think it's better if we don't pass the size here, and just strip the references in the implementation (unless there's reason to believe that there's a compilation time bottleneck here). `static constexpr size_t` variable seems like the most natural approach.

Reply:
I have no strong opinions here. What you say seems sensible. Modified this as suggested. Also adjusted `VisitImpl` in the same way.